src/components/global/Header/Header.tsx

```diff
@@ -0,0 +1,66 @@
+import React, {
+  FunctionComponent,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from "react";
+import { BookSearch } from "../../BookFinder/BookSearch/BookSearch";
+import axios from "axios";
+import { searchUrl } from "../../../helpers/BookHelper";
+import { notification } from "antd";
+import { searchContext } from "../../../contexts/SearchContext/SearchContext";
+import { useHistory, useLocation } from "react-router-dom";
+import styles from "./Header.module.scss";
+
+export const Header: FunctionComponent = () => {
+  const [query, setQuery] = useState("");
+  const history = useHistory();
+  const location = useLocation();
+  const { search, setSearch } = useContext(searchContext);
+
+  const onSearch = (queryTerm: string) => {
+    const pathname = location.pathname;
+    if (pathname !== "/") {
+      history.push("/");
+    }
+    setSearch({ ...search, startIndex: 0 });
+    setQuery(queryTerm);
+  };
+
+  const fetchBooks = useCallback(async () => {
+    setSearch({ ...search, isLoading: true });
+
+    try {
+      const result = await axios(
+        `${searchUrl}${query}&startIndex=${search.startIndex}`
+      );
+
+      setSearch({
+        ...search,
+        books: result.data.items,
+        total: result.data.totalItems,
+        isLoading: false,
+      });
+    } catch (e) {
+      notification.error({
+        message: "An error occurred",
+        description: "Please try searching again",
+      });
+    }
+
+    window.scrollTo(0, 0);
+  }, [query, search.startIndex, setSearch]);
```

Comment:
@JuraJuki ovde mi uporno govori da stavim search, a ne search.startIndex. ak stavim cijeli search objekt, onda mi se ova funkcija ne prestane pozivat i crasha mi app
pretpostavljam da zato kaj nije isti objekt pa ga to jebe il nekaj?
da to onda ignoriram il ima neki solution?

Reply:
Yup, mislim da Juki dobro govori za re-rendering. Što god koristiš unutar hooka moraš staviti u dependency array. Ako ti se search stalno mijenja po referenci, re-renderat će hook. Eslint rule react-hooks/exhaustive-deps provjerava to, mislim da bi ti trebao javiti to u stormu ako je kroz CRA napravljen projekt

useMemo bi to mogao fiksati, no ako imas kompleskniju promjenu unutar search objekta, onda ce ti mozda trebati ovo https://github.com/sandiiarov/use-deep-compare

Reply:
Svakako bi u dep arrayu trebalo ici [query, search, setSearch].

Reply:
dosel sam do istog rjesenja ko i ti reno, kad smo gledali skupa. s onom deep memoizacijom se vise nije dogadal loop, ali se odvrtil jos jedan ekstra request nakon kaj je prvi zavrsil.
cini mi se da zato jer sam u dependecyu imal memoizedSearch koji se updateal nakon prvog requesta (s novim vrijednostima), a kad se updatea memoizedSearch, ponovo se okinul useEffect i fetchBooks. napravil se novi get request, ali s obzirom da su isti podaci bili unutra (i isti objekt zbog memoizacije), nis se nije dogodilo. to mi se nije svidalo jer potencijalno more puknut taj drugi request, a nepotrebno se pozove
probal sam dodat onda provjeru isLoading u useEffect, al ni to nije pomoglo jer sam ga moral dodat u dependecy array pa sam se opet vratil na isto.

odlucil sam onda napravit context za svaku vrijednost tak da ne moram pozivat ...search ili ...memoizedSearch i to mi radi okej. jel vam se cini da je to malo overikill il je okej?

Reply:
Najbolje bi mozda bilo da taj call ne radis uopce na useEffect, nego bas na sam search. Mislim da ce olaksati taj handling state-a

Reply:
useEffect koristim jer mi se treba i na promjenu startIndexa okinut (zbog paginacije) :D
jel imas ideju kak da u tom slucaju avoidam useEffect?

Reply:
StartIndex se moze proslijediti u useCallback. Problem je što useEffect trenutno obuhvaća i akciju klika na search i inicijalnog loada. To razdvojiti logički mislim da će bit lakše handlat

Reply:
uspel sam napravit ovo s useAxiosom i useDeepMemo :D
napravil sam 2 useEffecta, jedan koji dohvati knjige kad se query il paginacija triggeraju, a drugi kad se data od useAxiosa promijeni. s tim drugim samo postavim rezultate u context koji je sad na jednom mjestu kak je i prije bil.

i sad nemam nikakvih linter errora i nema nikakvih ekstra callova!

Reply:
Nice! To je to ✅